### PR TITLE
test-bot: make master builds faster.

### PR DESCRIPTION
**Pull request by mikemcquaid**
Just do a few things rather than doing e.g. a full compile.

CC @xu-cheng @DomT4 who I'd talked about this with.
CC https://github.com/Homebrew/homebrew/issues/44320

### CODE REVIEW

File: Library/Homebrew/cmd/test-bot.rb [hunk 1]
```diff
@@ -437,10 +438,35 @@ def formula(formula_name)
...
+      if ARGV.include? "--fast"
+        formula_fetch_options = []
+        formula_fetch_options << "--force" if ARGV.include? "--cleanup"
+        formula_fetch_options << canonical_formula_name
+        test "brew", "fetch", "--retry", "--build-bottle",
+             *formula_fetch_options
+
+        if formula.bottled?
+          test "brew", "fetch", "--retry", *formula_fetch_options
+          install_args = ["--verbose", canonical_formula_name]
```
Comment (bfontaine) [inline Library/Homebrew/cmd/test-bot.rb]: Any reason to define this as a variable even if it’s used only once?

Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: Nope! It was used twice in an earlier iteration. I'll fix when I pull (to avoid another build)

File: Library/Homebrew/cmd/test-bot.rb [hunk 2]
```diff
@@ -437,10 +438,35 @@ def formula(formula_name)
         formula_name
       end
 
-      test "brew", "uses", canonical_formula_name
-
       formula = Formulary.factory(canonical_formula_name)
 
+      if ARGV.include? "--fast"
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/test-bot.rb]: Can we put all of this in a standalone method like `fast_formula_test` or something and use it in [here](https://github.com/mikemcquaid/homebrew/blob/test-bot-master-fast/Library/Homebrew/cmd/test-bot.rb#L765-L767) like:
```ruby
if ARGV.include? "--fast"
  formulae.each { |f| fast_formula_test(f) }
else
  formulae.each { |f| formula(f) }
end
```

Personally, I found long method with early return on different function is hard to follow.

**Comment by xu-cheng** [comment]
There are two thing you need to handle even in fast test.
* Handle the untapped deps
* Handle the conflict formulae

**Comment by mikemcquaid** [comment]
> There are two thing you need to handle even in fast test.

I think I'll just change the approach then so some steps are skipped on a fast test.

**Comment by mikemcquaid** [comment]
Have changed the approach so `--fast` just excludes some stuff instead.

File: Library/Homebrew/cmd/test-bot.rb [hunk 3]
```diff
@@ -883,7 +894,7 @@ def sanitize_ARGV_and_ENV
       ARGV << "--junit" << "--local"
     end
     if ARGV.include? "--ci-master"
-      ARGV << "--no-bottle" << "--email"
+      ARGV << "--no-bottle" << "--email" << "--fast"
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/test-bot.rb]: Do we still need `--no-bottle` flag? seems to be replaced by `--fast`.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: Good cal, getting rid of it.

File: Library/Homebrew/cmd/test-bot.rb [hunk 4]
```diff
@@ -547,35 +547,41 @@ def formula(formula_name)
...
-        test "brew", "install", *install_args
+        if !ARGV.include?("--fast") || formula_bottled
+          test "brew", "install", "--only-dependencies", *install_args unless dependencies.empty?
+          test "brew", "install", *install_args
+          install_passed = steps.last.passed?
+        else
+          install_passed = false
+        end
       end
-      install_passed = steps.last.passed?
```
Comment (bfontaine) [inline Library/Homebrew/cmd/test-bot.rb]: Just caught this: moving `install_passed` inside the `run_as_not_developer` block broke the `if install_passed` check; the variable is not defined anymore outside of the block. Fixed in a6611f9f.

Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: Whoops, thanks.
